runtime/autoload/man.vim

```diff
@@ -1,137 +1,290 @@
-    if winnr() > 1
-      exec thiswin . 'wincmd w'
-      while 1
-        if &filetype ==# 'man'
-          break
-        endif
-        wincmd w
-        if thiswin == winnr()
-          break
-        endif
-      endwhile
+" attempt to extract the name and sect out of 'name(sect)'
+" otherwise just return the largest string of valid characters in ref
+function! s:extract_sect_and_name_ref(ref) abort
+  if a:ref[0] ==# '-' " try ':Man -pandoc' with this disabled.
+    throw 'manpage name starts with ''-'''
+  endif
+  let ref = matchstr(a:ref, '[^()]\+([^()]\+)')
+  if empty(ref)
+    let name = matchstr(a:ref, '[^()]\+')
+    if empty(name)
+      throw 'manpage reference contains only parantheses'
     endif
+    return ['', name]
+  endif
+  let left = split(ref, '(')
+  " see ':Man 3X curses' on why tolower.
+  " TODO(nhooyr) Not sure if this is portable across OSs
+  " but I have not seen a single uppercase section.
+  return [tolower(split(left[1], ')')[0]), left[0]]
+endfunction
 
-    if &filetype !=# 'man'
-      let editcmd = 'tabnew'
+function! s:verify_exists(sect, name) abort
+  let path = system(s:man_cmd.' '.s:man_find_arg.' '.s:man_args(a:sect, a:name))
+  if path !~# '^\/'
+    if empty(a:sect)
+      throw 'no manual entry for '.a:name
+    endif
+    let path = system(s:man_cmd.' '.s:man_find_arg.' '.shellescape(a:name))
+    if path !~# '^\/'
+      throw 'no manual entry for '.a:name.'('.a:sect.') or '.a:name
     endif
   endif
+  call s:push_tag()
+  if a:name =~# '\/'
+    " We do not need to extract the section/name from the path if the name is
+    " just a path.
+    return ['', a:name]
+  endif
+  " We need to extract the section from the path because sometimes
+  " the actual section of the manpage is more specific than the section
+  " we provided to `man`. Try ':Man 3 App::CLI'.
+  " Also on linux, it seems that the name is case insensitive. So if one does
+  " ':Man PRIntf', we still want the name of the buffer to be 'printf' or
+  " whatever the correct capitilization is.
+  return s:extract_sect_and_name_path(path[:len(path)-2])
+endfunction
+
+let s:tag_stack = []
+
+function! s:push_tag() abort
+  let s:tag_stack += [{
+        \ 'buf':  bufnr('%'),
+        \ 'lnum': line('.'),
+        \ 'col':  col('.'),
+        \ }]
+endfunction
+
+function! man#pop_tag() abort
+  if !empty(s:tag_stack)
+    let tag = remove(s:tag_stack, -1)
+    execute tag['buf'].'b'
+    call cursor(tag['lnum'], tag['col'])
+  endif
+endfunction
 
-  silent exec editcmd.' man://'.page.(empty(sect)?'':'('.sect.')')
+" extracts the name and sect out of 'path/name.sect'
+function! s:extract_sect_and_name_path(path) abort
+  let tail = fnamemodify(a:path, ':t')
+  if a:path =~# '\.\%([glx]z\|bz2\|lzma\|Z\)$' " valid extensions
+    let tail = fnamemodify(tail, ':r')
+  endif
+  let sect = matchstr(tail, '\.\zs[^.]\+$')
+  let name = matchstr(tail, '^.\+\ze\.[^.]\+$')
+  return [sect, name]
+endfunction
+
+let s:manwidths = {}
```

Comment:
Why not use a `b:` variable (can be retrieved via `getbufvar()`) instead of re-inventing the `b:` dictionary here?